plutus-tx/src/PlutusTx/Builtins/Internal.hs

```diff
@@ -408,6 +408,11 @@ chooseList :: BuiltinList a -> b -> b -> b
 chooseList (BuiltinList [])    b1 _ = b1
 chooseList (BuiltinList (_:_)) _ b2 = b2
 
+{-# OPAQUE caseList' #-}
+caseList' :: forall a r . r -> (a -> BuiltinList a -> r) -> BuiltinList a -> r
```

Comment:
Why `caseList'` and not `caseList` (without the `'`)? Is there already a `caseList`?

Reply:
This is analogous to

```haskell
{-# INLINABLE matchList #-}
matchList :: forall a r . BI.BuiltinList a -> (() -> r) -> (a -> BI.BuiltinList a -> r) -> r
matchList l nilCase consCase = BI.caseList' nilCase (\x xs _ -> consCase x xs) l ()

{-# INLINABLE matchList' #-}
matchList' :: forall a r . BI.BuiltinList a -> r -> (a -> BI.BuiltinList a -> r) -> r
matchList' l nilCase consCase = BI.caseList' nilCase consCase l
```

but I should perhaps indeed add `caseList` for consistency. Thanks for bringing it up!

Reply:
Added `caseList` and `caseData`.

Reply:
`caseList` and `caseList'` are in different modules so it's fine calling both `caseList`, as in `Builtins.equalsInteger` and `Builtins.Internal.equalsInteger`.

Reply:
No, `caseList` and `caseList'` are different functions with different types, the same way `matchList` and `matchList'` (living in the same module) are different:

```haskell
caseList :: forall a r . (() -> r) -> (a -> BI.BuiltinList a -> r) -> BI.BuiltinList a -> r
```

vs

```haskell
caseList' :: forall a r . r -> (a -> BuiltinList a -> r) -> BuiltinList a -> r
```

hence we call them differently.